### ISSUE TITLE
[consistent-type-assertions] False positive with `allow-as-parameter`

### ISSUE THREAD

**Issue by fregante** [typescript-eslint/typescript-eslint#4462]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

https://typescript-eslint.io/play/#ts=4.5.2&sourceType=module&code=JYOwLgpgTgZghgYwgAgPJQOZxMAXnMYAexGQG8AoZa5EOAWwgH4AuZAZzClAwG4KAvhQoIS7IgBsIAOglEMACgDaZAcjjs0mbHgLEQAXQCUvIA&rules=N4IgAgLgngDgpgZwMYCcCWMIFpEBs0B2EA9EgPYEJoIRxFbTxYCGCCcKEaFCIAXAG0QHFGRQgANKFbtO3AgGVouOPxCtJIMgCMAVnCQQAMmlopmuACqw4AQTYcuPNRdxkA7iwRYYzcwFs4MxAAXwBdEKA&tsConfig=N4XyA

```JSON
{
  "rules": {
    "@typescript-eslint/consistent-type-assertions": [
      "error",
      {
        "assertionStyle": "as",
        "objectLiteralTypeAssertions": "allow-as-parameter"
      }
    ]
  }
}
```

```TS
interface Organization {
    name?: string;
}

console.log([{} as Organization]);
```

Note: The code is allowed if I drop the `[]` around it or if I drop `"allow-as-parameter"` from the config.

**Expected Result**

No error

**Actual Result**

```
Always prefer const x: T = { ... }.ESLint(@typescript-eslint/consistent-type-assertions)
```

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.10.0` |
| `@typescript-eslint/parser`        | `5.10.0` |
| `TypeScript`                       | `4.5.4` |
| `ESLint`                           | `8.1.0` |
| `node`                             | n/a |

**Comment by fregante** [comment]
Side note, this is allowed, but it feels like my example should be allowed too:

```ts
console.log([{}] as [Organization]);
```

**Comment by bradzacher** [comment]
The difficult thing behind both of these cases are that they are once removed from being an actual parameter - which makes them technically not a parameter any more.
The second example has a function boundary in between the object an the parameter - so IMO it is correct in its message - it's not in a parameter position, it's in a return-type position.

However the first is a little bit more nebulous.
`[{}] as T` is allowed because the rule does not check array expressions. This is because an [object expression asserted via an array expression is not unsafe, unlike `{} as T`](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgPJQOZxMAXnMYAexGQG8BYAKGVuRDgFsIB+ALmQGcwpQMBuagF9qoqgHpxyAKoA5AMoBBAGIBRaghLdkAD2QBecshhEiHAOTnkQ5HE5pM2PAWIhBEqUrUatYZAE8DZABtMmNTCyshAF1be2D0LBx8QhJo-iA).

But the bigger thing is "what does it mean to be a parameter". Currently the rule is coded that it must be a direct parameter. I could see an argument for it being an indirect parameter, however. The issue is that it's a bit of a can of worms as in "how deep can you go"?
I.e. if `[{} as T]` is okay, shouldn't `[[{} as T]]` be okay in a parameter position?
Then there's questions like "what about other constructs?" like an object. Should `{a: {} as T}` be okay in a parameter position? (and same question about the "depth")

(TBH I hate `allow-as-parameter` as it's really just a convenience option that trades off a variable declaration in favour of unsafe code.)

I don't think there's really a "correct" answer here. Given that this option has existed for ~3 years (#87) without other people running into this, I think that the behaviour is probably the "least surprising" for people.

---

The error message is probably the bigger problem here (suggesting `const x: T = {...}` when it's obviously invalid).

**Comment by fregante** [comment]
I'm using this plugin via a shared config so I don't know exactly what the expectations around the options are, but the issue remains: It's reporting on code that doesn't have alternatives. This leads my team to have to regularly silence the rule and losing confidence in it.

I'm not sure I have a solution either since assertions are indeed a smell to me, but as far as I know they're the only way to use nominal types.

**Comment by fregante** [comment]
> This is because an [object expression asserted via an array expression is not unsafe, unlike `{} as T`](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgPJQOZxMAXnMYAexGQG8BYAKGVuRDgFsIB+ALmQGcwpQMBuagF9qoqgHpxyAKoA5AMoBBAGIBRaghLdkAD2QBecshhEiHAOTnkQ5HE5pM2PAWIhBEqUrUatYZAE8DZABtMmNTCyshAF1be2D0LBx8QhJo-iA).

That's the kind of stuff I never want to see 😰 how is that not a TS bug…

**Comment by sindresorhus** [comment]
I would limit it to array and only one level. That would solve most cases.

**Comment by fregante** [comment]
I found another exception:

```ts
const withExtensionContext: ExtensionState = {
  isPending: false,
  manifest: {} as Manifest.ManifestBase, // <-- lint error
  openOptions: action("onOpenOptions"),
};
```

here I'm using the assertion specifically to override the type that I want the rest of the object to follow.

**Comment by bradzacher** [comment]
That is not a false positive and is completely correct. This is not a parameter position.
It is unsafe to use an `as` assertion in that location.

**Comment by fregante** [comment]
As far as I'm concerned, every `as` usage is unsafe, so if that's the intent then let's block *all* `as` usage. Since you probably wouldn't go as far as doing that, I believe that this usage is fine.

**Comment by bradzacher** [comment]
For reference - in flow all assertions must be type safe. So you really never use an assertion unless you're doing something unsafe.

It means some unnecessary checks in your code, but it for sure means you never do anything dumb like TS allows! I really don't mind it the vast majority of the time. A few edge cases but nothing that a suppression comment or cast through `any` can't work around easily.

So it is no surprise that I'm strongly in favour of never using assertions in TS code as well.
I would do it in this codebase as well - except the TS team insisted on making AST types that *rely* on assertions... 😢

---

The issue you're presenting is "when is an assertion safe and when is it unsafe?"

It would be possible to check that with a type-aware lint rule (though it would be VERY complex to check for), but without types really the safest stance to take is that they never ever are.

Because that could be overly restrictive with TS being somewhat dubious at times. So instead of disallowing all, instead "assertions on object literals are always unsafe" is more realistic and almost always true due to how the structural typing model is often times not as expected.

The `allow-as-parameter` option was added really only to help work around bad library types (https://github.com/typescript-eslint/typescript-eslint/issues/66). Extending the option's allowance to allow assertions within object literals is a significant expansion in scope and I don't think it's correct.

**Comment by bradzacher** [comment]
> The issue you're presenting is "when is an assertion safe and when is it unsafe?"
> It would be possible to check that with a type-aware lint rule 

We now have the [`no-unsafe-type-assertion`](https://typescript-eslint.io/rules/no-unsafe-type-assertion) (by "now" I mean Nov 2024).

I'd definitely say that the better course of action is allow all `as` assertions and rely on that rule instead.

Reviewing this issue from the lens of the current state of the project. This issue has 0 reactions and 2 participants. In 4 years it has seen no new activity from anyone. This fails our requirement "this proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal)." as it indicates that this is not a proposal that the broader community has interest in.

For this reason I'm going to close this off.
